specs/Picker-API_spec.md

```diff
@@ -0,0 +1,121 @@
+APIs in the Windows SDK. The Reunion Picker API will supercede the existing APIs and add additional methods
+for the functionality: File and folder picker, Multiple folder picker.
+
+- Windows.Storage.Pickers.FileOpenPicker (https://docs.microsoft.com/en-us/uwp/api/Windows.Storage.Pickers.FileOpenPicker)
+- Windows.Storage.Pickers.FileSavePicker (https://docs.microsoft.com/en-us/uwp/api/windows.storage.pickers.filesavepicker)
+- Windows.Storage.Pickers.FolderPicker (https://docs.microsoft.com/en-us/uwp/api/windows.storage.pickers.folderpicker) 
+
+# Description
+
+The goal of this API is to address the gaps in the existing File/Folder picker APIs. The API surface is similar to the existing
+Picker APIs in Windows SDK with additional methods to support File + Folder picker and multiple folder picker. The Project Reunion APIs will supercede the Windows APIs.
+
+# Examples
+## Show file and folder picker
+```c#
+    var picker = new StorageItemPicker();
+    picker.FiltersAndDescriptions.Insert("Images", "*.jpg, *.png");
+    picker.StartLocation = startLocation;
+    picker.PickerKinds = StorageItemPickerKinds.Folder | StorageItemPickerKinds.FileOpen;
+    // multi-select will be automatically set
+    IReadOnlyList<Windows.Storage.IStorageItem> storageItems = await picker.ShowPickerAsync();
+
+    if(0 < storageItems.Count)
+    {
+    }
+```
+## Show single folder picker
+```c#
+    var picker = new StorageItemPicker();
+    picker.PickerKinds = StorageItemPickerKinds.Folder;
+    IReadOnlyList<Windows.Storage.IStorageItem> folder = await picker.ShowPickerAsync();
+```
+## Show multi folder picker
+```c#
+    var picker = new StorageItemPicker();
+    picker.PickerKinds = StorageItemPickerKinds.Folder;
+    picker.MultiSelect = true;
+    IReadOnlyList<Windows.Storage.IStorageItem> folders = await picker.ShowPickerAsync();
+```
+## Show multiple file picker
+```c#
+    var picker = new StorageItemPicker();
+    picker.StartLocation = startLocation;
+    picker.MultuSelect = true;
+    picker.PickerKinds = StorageItemPickerKinds.FileOpen;
+    IReadOnlyList<Windows.Storage.IStorageItem> files = await picker.ShowPickerAsync();
+```
+## Show File Save picker
+```c#
+    var picker = new StorageItemPicker();
+    picker.PickerKinds = StorageItemPickerKinds.FileSave;
+    IReadOnlyList<Windows.Storage.IStorageItem> fileToSave = await picker.ShowPickerAsync();
+```
+
+# Remarks
+# API Notes
+# API Details
+```c#
+    namespace Microsoft.Reunion.Storage.Pickers
+    {
+        enum PickerViewMode
+        {
+            List = 0,
+            Details,
+            Tiles,
+            Content
+        };
+
+        enum StorageItemPickerKinds
+        {
+            FileOpen = 0x1,
+            Folder = 0x2,
+            FileSave = 0x4
+        };
+
+        runtimeclass StorageItemPicker
+        {
+            StorageItemPicker();
+
+            /// Properties
+            String StartLocation;
```

Comment:
This should probably be StartLocationFolderPath, since it's a path.  What happens if the app does not have access to this path? Is there a prompt for the user to approve picking from here to start with? Or is the act of picking implicit approval?

Reply:
IMO, this API app should accept an arbitrary folder path when running in an AppContainer, but not allow any other operations on that folder or its contents except through the usual methods.

Reply:
Can I specify a StartLocation via a StorageFolder e.g.
```
var sip = new StorageItemPicker();
sip.Start = Windows.Storage.KnownFolders.PicturesLibrary;
```

I can see merit in both
```
String StartFolderPath;
StorageFolder StartFolderLocation;
```
and `.ShowPickerAsync()` throws an exception if both are specified

Reply:
Woudn't you be able to just use the [Path](https://docs.microsoft.com/en-us/uwp/api/windows.storage.storagefolder.path?view=winrt-19041#Windows_Storage_StorageFolder_Path) API on the StorageFolder? E.g.:
```cs
var sip = new StorageItemPicker();
sip.StartFolderPath = Windows.Storage.KnownFolders.PicturesLibrary.Path;
```

The documentation for Path states "The full path of the current folder in the file system, if the path is available", so perhaps that's a scenario which would require one can specify a StorageFolder as the StartLocation (no full path available)?

Reply:
`StorageFolders` that represent libraries like the `PicturesLibrary` don't _necessarily_ have paths, since they are shell constructs and not filesystem directories. You could get the 'best' directory from [`SHGetKnownFolderPath`](https://docs.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetknownfolderpath) though (.NET has [an equivalent](https://docs.microsoft.com/en-us/dotnet/api/system.environment.getfolderpath?view=netcore-3.1)).

Reply:
Yes, as I have mentioned in the original thead [here](https://github.com/microsoft/ProjectReunion/issues/88#issuecomment-647758005), using the .NET equivalent could look like:
```cs
var filePicker = new FileOpenPicker()
{
    SuggestedStartLocation = Environment.GetFolderPath(Environment.SpecialFolder.MyPictures)
};
```
so at least in that case, we would be fine without shipping a specific `StorageFolder StartFolderLocation` API.

Reply:
> IMO, this API app should accept an arbitrary folder path when running in an AppContainer, but not allow any other operations on that folder or its contents except through the usual methods.

Yes, the point of the picker is to get files you don't have access to. If the user doesn't like the location, they can change it. 

Relying on a single, fixed list of locations (current `Windows.` API) or relying on the app already having access vastly reduces the functionality. For example, if I already have access to `C:\Foo\bar.txt` I might want the user to pick another file from `C:\Foo\` even though I don't have access. (I believe the current API remembers the last start location, but that doesn't help if the user picked another file in the meantime...)/

Reply:
The new pickers should be easy to nest into user controls or dialogs, and should allow simple access from MVVM viewmodels. Don't repeat WPF's mistakes.